File: ecommerce-monorepo/apps/invoice/src/services/pdf-generator.ts
import PDFDocument from "pdfkit";

export interface InvoiceItem {
  name: string;
  description?: string;
  quantity: number;
  price: number;
  tax?: number;
  discount?: number;
}

export interface Address {
  street: string;
  city: string;
  state: string;
  zipCode: string;
  country?: string;
}

export interface CompanyInfo {
  name: string;
  logo?: string;
  address: Address;
  phone?: string;
  email?: string;
  website?: string;
  taxId?: string;
}

export interface InvoiceData {
  orderId: string;
  date: string;
  dueDate?: string;
  customerName: string;
  customerEmail: string;
  customerPhone?: string;
  totalAmount: number;
  subtotal?: number;
  tax?: number;
  discount?: number;
  shippingCost?: number;
  items: InvoiceItem[];
  address: Address;
  notes?: string;
  paymentMethod?: string;
  paymentStatus?: "Paid" | "Pending" | "Overdue";
  currency?: string;
}

export interface InvoiceOptions {
  company?: CompanyInfo;
  colors?: {
    primary: string;
    secondary: string;
    text: string;
    lightGray: string;
  };
}

export class PdfGenerator {
  private static readonly DEFAULT_COMPANY: CompanyInfo = {
    name: "E-Commerce Co.",
    address: {
      street: "123 Cloud Avenue",
      city: "Tech City",
      state: "TC",
      zipCode: "90210",
      country: "USA",
    },
    phone: "+1 (555) 123-4567",
    email: "support@ecommerce.com",
    website: "www.ecommerce.com",
    taxId: "TAX-123456789",
  };

  private static readonly DEFAULT_COLORS = {
    primary: "#9333ea",
    secondary: "#c084fc",
    text: "#1f2937",
    lightGray: "#f3f4f6",
  };

  static async generate(
    data: InvoiceData,
    options: InvoiceOptions = {}
  ): Promise<Buffer> {
    return new Promise((resolve, reject) => {
      const doc = new PDFDocument({ size: "A4", margin: 50 });
      const chunks: Buffer[] = [];

      doc.on("data", (chunk) => chunks.push(chunk));
      doc.on("end", () => resolve(Buffer.concat(chunks)));
      doc.on("error", (err) => reject(err));

      const company = options.company || this.DEFAULT_COMPANY;
      const colors = options.colors || this.DEFAULT_COLORS;
      const currency = data.currency || "$";

      // Header
      this.generateHeader(doc, company, colors);

      // Invoice Info & Customer Details
      this.generateInvoiceInfo(doc, data, colors);

      // Items Table
      this.generateItemsTable(doc, data, colors, currency);

      // Totals
      this.generateTotals(doc, data, colors, currency);

      // Notes & Payment Info
      this.generateFooter(doc, data, company, colors);

      doc.end();
    });
  }

  private static generateHeader(
    doc: PDFKit.PDFDocument,
    company: CompanyInfo,
    colors: any
  ): void {
    // Company name and logo area
    doc
      .fillColor(colors.primary)
      .fontSize(28)
      .font("Helvetica-Bold")
      .text(company.name, 50, 50);

    // Invoice title on the right
    doc
      .fillColor(colors.primary)
      .fontSize(32)
      .font("Helvetica-Bold")
      .text("INVOICE", 400, 50, { align: "right" });

    // Company details
    doc
      .fillColor(colors.text)
      .fontSize(9)
      .font("Helvetica")
      .text(company.address.street, 50, 85)
      .text(
        `${company.address.city}, ${company.address.state} ${company.address.zipCode}`,
        50,
        97
      );

    if (company.address.country) {
      doc.text(company.address.country, 50, 109);
    }

    if (company.phone) {
      doc.text(`Phone: ${company.phone}`, 50, 121);
    }

    if (company.email) {
      doc.text(`Email: ${company.email}`, 50, 133);
    }

    if (company.taxId) {
      doc.text(`Tax ID: ${company.taxId}`, 50, 145);
    }

    doc.moveDown(3);
  }

  private static generateInvoiceInfo(
    doc: PDFKit.PDFDocument,
    data: InvoiceData,
    colors: any
  ): void {
    const startY = 190;

    // Invoice details box (left)
    doc.fillColor(colors.lightGray).rect(50, startY, 240, 100).fill();

    doc
      .fillColor(colors.primary)
      .fontSize(10)
      .font("Helvetica-Bold")
      .text("Invoice Details", 60, startY + 10);

    doc
      .fillColor(colors.text)
      .fontSize(9)
      .font("Helvetica")
      .text(`Invoice Number:`, 60, startY + 30)
      .font("Helvetica-Bold")
      .text(data.orderId.toUpperCase(), 130, startY + 30)
      .font("Helvetica")
      .text(`Invoice Date:`, 60, startY + 45)
      .font("Helvetica-Bold")
      .text(data.date, 130, startY + 45);

    if (data.dueDate) {
      doc
        .font("Helvetica")
        .text(`Due Date:`, 60, startY + 60)
        .font("Helvetica-Bold")
        .text(data.dueDate, 150, startY + 60);
    }

    if (data.paymentStatus) {
      const statusColor =
        data.paymentStatus === "Paid"
          ? "#10b981"
          : data.paymentStatus === "Pending"
            ? "#f59e0b"
            : "#ef4444";

      doc
        .font("Helvetica")
        .text(`Status:`, 60, startY + 75)
        .fillColor(statusColor)
        .font("Helvetica-Bold")
        .text(data.paymentStatus, 150, startY + 75);
    }

    // Customer details box (right)
    doc.fillColor(colors.lightGray).rect(310, startY, 240, 100).fill();

    doc
      .fillColor(colors.primary)
      .fontSize(10)
      .font("Helvetica-Bold")
      .text("Bill To", 320, startY + 10);

    doc
      .fillColor(colors.text)
      .fontSize(9)
      .font("Helvetica-Bold")
      .text(data.customerName, 320, startY + 30)
      .font("Helvetica")
      .text(data.customerEmail, 320, startY + 45);

    if (data.customerPhone) {
      doc.text(data.customerPhone, 320, startY + 58);
    }

    doc
      .text(data.address.street, 320, startY + 71)
      .text(
        `${data.address.city}, ${data.address.state} ${data.address.zipCode}`,
        320,
        startY + 83
      );

    doc.moveDown(4);
  }

  private static generateItemsTable(
    doc: PDFKit.PDFDocument,
    data: InvoiceData,
    colors: any,
    currency: string
  ): void {
    const tableTop = 330;
    const itemHeight = 30;

    // Table header
    doc.fillColor(colors.primary).rect(50, tableTop, 500, 25).fill();

    doc
      .fillColor("#ffffff")
      .fontSize(10)
      .font("Helvetica-Bold")
      .text("Description", 60, tableTop + 8)
      .text("Qty", 340, tableTop + 8)
      .text("Price", 400, tableTop + 8)
      .text("Amount", 480, tableTop + 8);

    // Table rows
    let currentY = tableTop + 25;

    data.items.forEach((item, index) => {
      const isEven = index % 2 === 0;
      const lineTotal = item.price * item.quantity;

      // Alternating row background
      if (isEven) {
        doc
          .fillColor(colors.lightGray)
          .rect(50, currentY, 500, itemHeight)
          .fill();
      }

      doc
        .fillColor(colors.text)
        .fontSize(10)
        .font("Helvetica-Bold")
        .text(item.name, 60, currentY + 8, { width: 260 });

      if (item.description) {
        doc
          .fontSize(8)
          .font("Helvetica")
          .fillColor("#6b7280")
          .text(item.description, 60, currentY + 20, { width: 260 });
      }

      doc
        .fillColor(colors.text)
        .fontSize(10)
        .font("Helvetica")
        .text(item.quantity.toString(), 340, currentY + 8)
        .text(`${currency}${item.price.toFixed(2)}`, 400, currentY + 8)
        .font("Helvetica-Bold")
        .text(`${currency}${lineTotal.toFixed(2)}`, 480, currentY + 8);

      currentY += itemHeight;
    });

    doc.moveDown(2);
  }

  private static generateTotals(
    doc: PDFKit.PDFDocument,
    data: InvoiceData,
    colors: any,
    currency: string
  ): void {
    const rightAlign = 500; // Aligned with the right edge of standard content
    const labelAlign = 350; // Where "Subtotal:" starts
    let currentY = doc.y + 20;

    // Helper to safely format currency (handles Strings and Numbers)
    const formatMoney = (amount: number | string | undefined) => {
      const num = Number(amount || 0);
      return `${currency}${num.toFixed(2)}`;
    };

    doc.fontSize(10).font("Helvetica");

    // 1. Subtotal
    if (data.subtotal) {
      doc
        .fillColor(colors.text)
        .text("Subtotal:", labelAlign, currentY)
        .text(formatMoney(data.subtotal), rightAlign, currentY, {
          align: "right",
        });
      currentY += 20;
    }

    // 2. Discount (Green text if applied)
    if (data.discount && Number(data.discount) > 0) {
      doc
        .fillColor("#10b981") // Green
        .text("Discount:", labelAlign, currentY)
        .text(`- ${formatMoney(data.discount)}`, rightAlign, currentY, {
          align: "right",
        });
      currentY += 20;
    }

    // 3. Tax
    if (data.tax && Number(data.tax) > 0) {
      doc
        .fillColor(colors.text)
        .text("Tax / VAT:", labelAlign, currentY)
        .text(formatMoney(data.tax), rightAlign, currentY, { align: "right" });
      currentY += 20;
    }

    // 4. Shipping
    if (data.shippingCost && Number(data.shippingCost) > 0) {
      doc
        .text("Shipping:", labelAlign, currentY)
        .text(formatMoney(data.shippingCost), rightAlign, currentY, {
          align: "right",
        });
      currentY += 20;
    }

    // --- DIVIDER LINE ---
    currentY += 5;
    doc
      .strokeColor("#e5e7eb")
      .lineWidth(1)
      .moveTo(labelAlign, currentY)
      .lineTo(rightAlign + 50, currentY) // Extend slightly past the numbers
      .stroke();
    currentY += 15;

    // 5. Total Amount (Large & Bold)
    doc.font("Helvetica-Bold").fontSize(14).fillColor(colors.primary);

    doc.text("Total Amount:", labelAlign, currentY, { width: 120 });
    doc
      .fontSize(16)
      .text(formatMoney(data.totalAmount), rightAlign, currentY - 2, {
        align: "right",
      });

    currentY += 35;

    // 6. Payment Status Badge (Optional but professional)
    if (data.paymentStatus) {
      const statusColor =
        data.paymentStatus === "Paid"
          ? "#10b981" // Green
          : data.paymentStatus === "Overdue"
            ? "#ef4444" // Red
            : "#f59e0b"; // Orange (Pending)

      doc.fontSize(10).font("Helvetica-Bold").fillColor(colors.text);
      doc.text("Payment Status:", labelAlign, currentY);

      doc
        .fillColor(statusColor)
        .text(data.paymentStatus.toUpperCase(), rightAlign, currentY, {
          align: "right",
        });
      currentY += 15;
    }

    // 7. Payment Method (e.g. "Credit Card ending in 4242")
    if (data.paymentMethod) {
      doc.font("Helvetica").fillColor(colors.text);
      doc.text("Payment Method:", labelAlign, currentY);
      doc.text(data.paymentMethod, rightAlign, currentY, { align: "right" });
    }
  }

  private static generateFooter(
    doc: PDFKit.PDFDocument,
    data: InvoiceData,
    company: CompanyInfo,
    colors: any
  ): void {
    let footerY = doc.y + 40;
    const amountWidth = 160;

    // Notes section
    if (data.notes) {
      doc
        .fillColor(colors.primary)
        .fontSize(10)
        .font("Helvetica-Bold")
        .text("Notes:", 50, footerY);

      doc
        .fillColor(colors.text)
        .fontSize(9)
        .font("Helvetica")
        .text(data.notes, 50, footerY + 15, { width: 500, align: "left" });

      footerY = doc.y + 20;
    }

    // Payment information
    if (data.paymentMethod) {
      doc
        .fillColor(colors.primary)
        .fontSize(10)
        .font("Helvetica-Bold")
        .text("Payment Method:", 50, footerY);

      doc
        .fillColor(colors.text)
        .fontSize(9)
        .font("Helvetica")
        .text(data.paymentMethod, 50, footerY + 15);

      footerY = doc.y + 20;
    }

    // Bottom section
    const bottomY = 750;
    doc
      .strokeColor(colors.lightGray)
      .lineWidth(1)
      .moveTo(50, bottomY)
      .lineTo(550, bottomY)
      .stroke();

    doc
      .fillColor("#6b7280")
      .fontSize(9)
      .font("Helvetica")
      .text("Thank you for your business!", 50, bottomY + 10, {
        align: "center",
      })
      .text(
        `For questions, contact: ${company.email || "support@company.com"}`,
        50,
        bottomY + 25,
        { align: "center" }
      );
  }
}

File: ecommerce-monorepo/apps/invoice/internal/invoice/pdf_generator.go
package invoice

import (
	"bytes"
	"fmt"
	"strings"
	"time"

	"github.com/jung-kurt/gofpdf"
	"github.com/tomarrohitt/invoice-go/internal/events"
)

type PDFGenerator struct {
	PrimaryColor []int
	TextColor    []int
	GrayColor    []int
}

func NewPDFGenerator() *PDFGenerator {
	return &PDFGenerator{
		PrimaryColor: []int{147, 51, 234},
		TextColor:    []int{31, 41, 55},
		GrayColor:    []int{243, 244, 246},
	}
}

func (g *PDFGenerator) Generate(event events.OrderPaidEvent, invoiceID string) ([]byte, error) {
	pdf := gofpdf.New("P", "mm", "A4", "")
	pdf.SetMargins(15, 15, 15)
	pdf.AddPage()

	g.generateHeader(pdf)
	g.generateInfoSection(pdf, event, invoiceID)
	g.generateTable(pdf, event)
	g.generateTotals(pdf, event)
	g.generateFooter(pdf)

	var buf bytes.Buffer
	if err := pdf.Output(&buf); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

func (g *PDFGenerator) generateHeader(pdf *gofpdf.Fpdf) {
	pdf.SetFont("Arial", "B", 24)
	pdf.SetTextColor(g.PrimaryColor[0], g.PrimaryColor[1], g.PrimaryColor[2])
	pdf.Cell(0, 10, "E-Commerce Co.")

	pdf.SetFont("Arial", "", 9)
	pdf.SetTextColor(g.TextColor[0], g.TextColor[1], g.TextColor[2])
	pdf.SetXY(120, 25)
	pdf.Cell(18, 5, "Invoice No:")
	pdf.SetXY(120, 30)
	pdf.Cell(18, 5, "Date:")

	pdf.SetXY(15, 25)
	pdf.SetFont("Arial", "", 9)
	pdf.SetTextColor(g.TextColor[0], g.TextColor[1], g.TextColor[2])
	pdf.Text(15, 30, "123 Cloud Avenue, Tech City")
	pdf.Text(15, 35, "Phone: +1 (555) 123-4567")
	pdf.Text(15, 40, "Email: support@ecommerce.com")
	pdf.Ln(20)
}

func (g *PDFGenerator) generateInfoSection(pdf *gofpdf.Fpdf, event events.OrderPaidEvent, invoiceID string) {

	shortID := fmt.Sprintf("%s-%s", invoiceID[:8], invoiceID[len(invoiceID)-12:])

	pdf.SetFont("Arial", "", 9)
	pdf.SetTextColor(g.TextColor[0], g.TextColor[1], g.TextColor[2])
	pdf.SetXY(140, 25)
	pdf.Cell(0, 5, fmt.Sprintf("INV-%s", strings.ToUpper(shortID)))
	pdf.SetXY(140, 30)
	pdf.Cell(0, 5, time.Now().Format("Jan 02, 2006"))

	pdf.SetFillColor(g.GrayColor[0], g.GrayColor[1], g.GrayColor[2])
	pdf.Rect(15, 55, 180, 30, "F")

	pdf.SetXY(20, 60)
	pdf.SetFont("Arial", "B", 10)
	pdf.SetTextColor(g.PrimaryColor[0], g.PrimaryColor[1], g.PrimaryColor[2])
	pdf.Cell(0, 5, "Bill To")

	pdf.SetFont("Arial", "B", 9)
	pdf.SetTextColor(g.TextColor[0], g.TextColor[1], g.TextColor[2])
	pdf.SetXY(20, 68)
	pdf.Cell(0, 5, event.Data.UserName)
	pdf.SetFont("Arial", "", 9)
	pdf.SetXY(20, 74)
	pdf.Cell(0, 5, event.Data.UserEmail)

	pdf.Rect(15, 90, 90, 30, "F")
	pdf.SetXY(20, 95)
	pdf.SetFont("Arial", "B", 10)
	pdf.SetTextColor(g.PrimaryColor[0], g.PrimaryColor[1], g.PrimaryColor[2])
	pdf.Cell(0, 5, "Shipping Address")

	pdf.SetFont("Arial", "", 8)
	pdf.SetTextColor(g.TextColor[0], g.TextColor[1], g.TextColor[2])
	pdf.SetXY(20, 102)
	pdf.Cell(0, 4, event.Data.ShippingAddress.Name)
	pdf.SetXY(20, 106)
	pdf.Cell(0, 4, event.Data.ShippingAddress.Street)
	pdf.SetXY(20, 110)
	pdf.Cell(0, 4, fmt.Sprintf("%s, %s %s", event.Data.ShippingAddress.City, event.Data.ShippingAddress.State, event.Data.ShippingAddress.ZipCode))
	pdf.SetXY(20, 114)
	pdf.Cell(0, 4, event.Data.ShippingAddress.Country)

	pdf.Rect(105, 90, 90, 30, "F")
	pdf.SetXY(110, 95)
	pdf.SetFont("Arial", "B", 10)
	pdf.SetTextColor(g.PrimaryColor[0], g.PrimaryColor[1], g.PrimaryColor[2])
	pdf.Cell(0, 5, "Billing Address")

	pdf.SetFont("Arial", "", 8)
	pdf.SetTextColor(g.TextColor[0], g.TextColor[1], g.TextColor[2])
	pdf.SetXY(110, 102)
	pdf.Cell(0, 4, event.Data.BillingAddress.Name)
	pdf.SetXY(110, 106)
	pdf.Cell(0, 4, event.Data.BillingAddress.Street)
	pdf.SetXY(110, 110)
	pdf.Cell(0, 4, fmt.Sprintf("%s, %s %s", event.Data.BillingAddress.City, event.Data.BillingAddress.State, event.Data.BillingAddress.ZipCode))
	pdf.SetXY(110, 114)
	pdf.Cell(0, 4, event.Data.BillingAddress.Country)
}

func (g *PDFGenerator) generateTable(pdf *gofpdf.Fpdf, event events.OrderPaidEvent) {
	pdf.SetXY(15, 125)
	pdf.SetFillColor(g.PrimaryColor[0], g.PrimaryColor[1], g.PrimaryColor[2])
	pdf.SetTextColor(255, 255, 255)
	pdf.SetFont("Arial", "B", 10)

	pdf.CellFormat(80, 10, "Description", "0", 0, "L", true, 0, "")
	pdf.CellFormat(30, 10, "Qty", "0", 0, "C", true, 0, "")
	pdf.CellFormat(40, 10, "Price", "0", 0, "C", true, 0, "")
	pdf.CellFormat(30, 10, "Amount", "", 1, "R", true, 0, "")

	pdf.SetTextColor(g.TextColor[0], g.TextColor[1], g.TextColor[2])
	pdf.SetFont("Arial", "", 10)
	for i, item := range event.Data.Items {
		fill := i%2 == 0
		if fill {
			pdf.SetFillColor(g.GrayColor[0], g.GrayColor[1], g.GrayColor[2])
		} else {
			pdf.SetFillColor(255, 255, 255)
		}

		lineTotal := item.Price * float64(item.Quantity)
		pdf.CellFormat(80, 10, item.Name, "0", 0, "L", true, 0, "")
		pdf.CellFormat(30, 10, fmt.Sprintf("%d", item.Quantity), "0", 0, "C", true, 0, "")
		pdf.CellFormat(40, 10, fmt.Sprintf("$%.2f", item.Price), "0", 0, "C", true, 0, "")
		pdf.CellFormat(30, 10, fmt.Sprintf("$%.2f", lineTotal), "", 1, "R", true, 0, "")
	}
}

func (g *PDFGenerator) generateTotals(pdf *gofpdf.Fpdf, event events.OrderPaidEvent) {
	pdf.Ln(2)

	var startX float64 = 140
	var labelWidth float64 = 35
	var valueWidth float64 = 0

	pdf.SetFont("Arial", "", 10)
	pdf.SetTextColor(g.TextColor[0], g.TextColor[1], g.TextColor[2])
	pdf.SetX(startX)
	pdf.CellFormat(labelWidth, 8, "Subtotal:", "", 0, "L", false, 0, "")
	pdf.CellFormat(valueWidth, 8, fmt.Sprintf("$%.2f", event.Data.Subtotal), "", 1, "R", false, 0, "")

	pdf.SetX(startX)
	pdf.CellFormat(labelWidth, 8, "Tax:", "", 0, "L", false, 0, "")
	pdf.CellFormat(valueWidth, 8, fmt.Sprintf("$%.2f", event.Data.TaxedAmount), "", 1, "R", false, 0, "")

	pdf.SetDrawColor(g.GrayColor[0], g.GrayColor[1], g.GrayColor[2])
	pdf.Line(startX, pdf.GetY(), startX+labelWidth+valueWidth, pdf.GetY())

	pdf.SetFont("Arial", "B", 10)
	pdf.SetTextColor(g.PrimaryColor[0], g.PrimaryColor[1], g.PrimaryColor[2])
	pdf.SetX(startX)
	pdf.CellFormat(labelWidth, 8, "Total Amount:", "", 0, "L", false, 0, "")
	pdf.SetFont("Arial", "B", 10)
	pdf.CellFormat(valueWidth, 8, fmt.Sprintf("$%.2f", event.Data.TotalAmount), "", 1, "R", false, 0, "")
}

func (g *PDFGenerator) generateFooter(pdf *gofpdf.Fpdf) {
	pdf.SetAutoPageBreak(false, 0)

	pdf.SetY(-25)

	pdf.SetDrawColor(g.GrayColor[0], g.GrayColor[1], g.GrayColor[2])
	pdf.SetLineWidth(0.2)
	pdf.Line(15, pdf.GetY(), 195, pdf.GetY())

	pdf.Ln(2)

	pdf.SetFont("Arial", "I", 8)
	pdf.SetTextColor(128, 128, 128)

	pdf.CellFormat(0, 5, "Thank you for your business!", "", 1, "C", false, 0, "")
	pdf.CellFormat(0, 5, "For questions, contact: support@ecommerce.com", "", 1, "C", false, 0, "")

	pdf.SetAutoPageBreak(true, 15)
}
